File: p020e/valid_parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];
  for (let c of s) {
    if (c === '(') stack.push(')');
    else if (c === '{') stack.push('}');
    else if (c === '[') stack.push(']');
    else {
      if (stack.length == 0 || stack.pop() !== c) return false;
    }
  }
  return stack.length === 0;
};
// TEST
[
  ['{}', true],
  ['()[]{}', true],
  ['(]', false],
  ['([)]', false],
  ['{[]}', true],
].forEach(([s, expected]) => {
  const actual = isValid(s);
  console.log('Is', s, 'a valid parentheses? ->', actual);
  console.assert(actual === expected);
});

File: p020e/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == "(":
                stack.append(")")
            elif c == "{":
                stack.append("}")
            elif c == "[":
                stack.append("]")
            else:
                if len(stack) == 0 or stack.pop() != c:
                    return False
        return len(stack) == 0


# TESTS
for s, expected in [
    ("{}", True),
    ("()[]{}", True),
    ("(]", False),
    ("([)]", False),
    ("{[]}", True),
]:
    sol = Solution()
    actual = sol.isValid(s)
    print("Is", s, "a valid parentheses? ->", actual)
    assert actual == expected
